lib/plugin-manager.js

```diff
@@ -0,0 +1,303 @@
+      return {
+        name: source.name,
+        type: source.type,
+        status: source.status()
+      };
+    });
+  }
+
+  /**
+   * Coordinates synchronizing the index and metadata sources and then mapping them together
+   * @private
+   */
+  _init() {
+    const boundIndex = this._index.bind(this);
+    const boundMapper = this._mapper.bind(this);
+
+    async.series([boundIndex, boundMapper], (err, response) => {
+      // There's no possible error condition because both this._mapper() and this._index() handle
+      // their own errors in order to retry their respective operations.
+      this._ok = true;
+
+      const sources = response[1];
+
+      this.emit('sources-generated', sources);
+      this._registerSources(sources);
+      this.emit('sources-registered', this.storage.sources);
+    });
+  }
+
+  /**
+   * Coordinate fetching the PluginManager's index and metadata source
+   * @param {Function} callback
+   * @returns {null}
+   * @private
+   */
+  _index(callback) {
+    return async.parallel([
+      this._createParallelTask(this.index),
+      this._createParallelTask(this.metadata)
+    ], (err) => {
+      if (err) {
+        // If any of the parallel tasks fail it means either the Metadata plugin or the index plugin has failed
+        // so we need to determine which plugin threw the error then set a listener on it for its next update to
+        // try to assemble the whole thing again.
+        this._error(err);
+
+        // ECONNREFUSED seems to be the only code that indicates an EC2MetadataService failure
+        if (err.code === 'ECONNREFUSED') {
+          this.metadata.shutdown();
+          this.metadata.once('update', this._boundInit);
+          this.metadata.initialize();
+        } else {
+          this.index.shutdown();
+          this.index.once('update', this._boundInit);
+          this.index.initialize();
+        }
+        return;
+      }
+      callback(null);
+    });
+  }
+
+  /**
+   * Maps instance metadata into the index document
+   * @param {Function} callback
+   * @private
+   */
+  _mapper(callback) {
+    let sources = [];
+
+    try {
+      sources = this.index.properties.sources.map((el) => {
+        return iter(el, (k, v) => {
+          // TODO: No handling for arrays. Are we expecting arrays anywhere in the index?
+          return StringTemplate.coerce(v, this.metadata.properties);
+        });
+      });
+      callback(null, sources);
+    } catch (mappingErr) {
+      this._error(mappingErr);
+      setTimeout(this._boundInit, this.updateDelay);
+    }
+  }
+
+  /**
+   * Create function that feeds async.parallel (or any async marshalling function)
+   * @param {Metadata|S3} source
+   * @returns {Function}
+   * @private
+   */
+  _createParallelTask(source) {
+    return (c) => {
+      const callBackErr = (err) => {
+        source.removeListener('update', callBackSuccess); // eslint-disable-line no-use-before-define
+        c(err);
```

Comment:
I think adding `source.type` to the callback is a great idea. I'm going to do some investigation to see if it's actually possible to pass additional arguments into the `err` case of `async.parallel`.